api/crawler/client.go

```diff
@@ -0,0 +1,332 @@
+package crawler
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+)
+
+const (
+	// DefaultBaseURL is the default base URL for the Algolia Crawler API.
+	DefaultBaseURL = "https://crawler.algolia.com/api/1/"
+)
+
+// Client provides methods to interact with the Algolia Crawler API.
+type Client struct {
+	UserID string
+	APIKey string
+
+	client *http.Client
+}
+
+// NewClient returns a new Crawler API client.
+func NewClient(userID, apiKey string) *Client {
+	return &Client{
+		UserID: userID,
+		APIKey: apiKey,
+		client: http.DefaultClient,
+	}
+}
+
+// NewClientWithHTTPClient returns a new Crawler API client with a custom HTTP client.
+func NewClientWithHTTPClient(userID, apiKey string, client *http.Client) *Client {
+	return &Client{
+		UserID: userID,
+		APIKey: apiKey,
+		client: client,
+	}
+}
+
+// Request sends an HTTP request and returns an HTTP response.
+// It unmarshals the response body to the given interface.
+func (c *Client) request(res interface{}, method string, path string, body interface{}, urlParams map[string]string) error {
+	r, err := c.buildRequest(method, path, body, urlParams)
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.client.Do(r)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode >= 400 {
+		var errResp ErrResponse
+		if err := unmarshalTo(resp, &errResp); err != nil {
+			return err
+		}
+
+		if errResp.Err.Errors != nil {
+			var errs []string
+			for _, e := range errResp.Err.Errors {
+				errs = append(errs, e.Message)
+			}
+			return fmt.Errorf("%s: %s", errResp.Err.Message, errs)
+		}
+
+		return errors.New(errResp.Err.Message)
+	}
+
+	if res != nil {
+		if err := unmarshalTo(resp, res); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// buildRequestWithoutBody builds an HTTP request without a body.
+func (c *Client) buildRequestWithoutBody(method, url string) (*http.Request, error) {
+	return http.NewRequest(method, url, nil)
+}
+
+// buildRequestWithBody builds an HTTP request with a body.
+func (c *Client) buildRequestWithBody(method, url string, body interface{}) (*http.Request, error) {
```

Comment:
Question here: since this function is in the crawler client context, do you think it'd be relevant to create a type for the body? 

```go
type Body struct{
	URLs []string `json:"urls"`
	Save bool     `json:"save"`
}
```

Reply:
For each of the specific method you mean? Honestly, since the methods only accept a small numbers of arguments, I don't any reason to expose the struct outside of the scope. If it was a long list of arguments, I would say it would be better to expose a struct externally indeed, but it's not the case here.